### DISCUSSION TITLE
Spacing between fields in completion menu

### DISCUSSION THREAD

**Discussion by emnnipal** [LazyVim/LazyVim#1221] [Q&A] [answered]
Hi, I would like to know how I can remove the space between each completion field? To provide better clarity, here's a screenshot

<img width="289" alt="Screenshot 2023-07-19 at 1 37 31 PM" src="https://github.com/LazyVim/LazyVim/assets/22177776/f2ec14ba-6cf0-46f4-8a80-feafec86eb55">

I would like to have the space removed.

Here's my config for nvim-cmp

```
{
    "hrsh7th/nvim-cmp",
    dependencies = {
      "hrsh7th/cmp-emoji",
    },
    ---@param opts cmp.ConfigSchema
    opts = function(_, opts)
      local has_words_before = function()
        unpack = unpack or table.unpack
        local line, col = unpack(vim.api.nvim_win_get_cursor(0))
        return col ~= 0 and vim.api.nvim_buf_get_lines(0, line - 1, line, true)[1]:sub(col, col):match("%s") == nil
      end

      local luasnip = require("luasnip")
      local cmp = require("cmp")

      opts.mapping = vim.tbl_extend("force", opts.mapping, {
        ["<Tab>"] = cmp.mapping(function(fallback)
          if cmp.visible() then
          elseif luasnip.expand_or_locally_jumpable() then
            luasnip.expand_or_jump()
          elseif has_words_before() then
            cmp.complete()
          else
            fallback()
          end
        end, { "i", "s" }),
        ["<S-Tab>"] = cmp.mapping(function(fallback)
          if cmp.visible() then
            cmp.select_prev_item()
          elseif luasnip.jumpable(-1) then
            luasnip.jump(-1)
          else
            fallback()
          end
        end, { "i", "s" }),
      })

      opts.window = {
        completion = cmp.config.window.bordered({
          side_padding = 0, -- to prevent icons from being cut off
        }),
        documentation = cmp.config.window.bordered(),
      }


      opts.formatting = {
        fields = { "kind", "abbr", "menu" }, -- order of fields
        format = function(entry, item)
          local icons = require("lazyvim.config").icons.kinds

          local max_width = 26
          local max_detail_width = 20

          local source_name = ({
            nvim_lsp = "(LSP)",
            emoji = "(Emoji)",
            path = "(Path)",
            calc = "(Calc)",
            cmp_tabnine = "(Tabnine)",
            vsnip = "(Snippet)",
            luasnip = "(Snippet)",
            buffer = "(Buffer)",
            tmux = "(TMUX)",
            copilot = "(Copilot)",
            treesitter = "(TreeSitter)",
          })[entry.source.name]

          local detail = string.sub(entry.completion_item.detail or "", 1, max_detail_width)
          item.menu = string.format("%s %s", source_name, detail)

          if icons[item.kind] then
            item.kind = icons[item.kind]
          end

          if max_width ~= 0 and #item.abbr > max_width then
            item.abbr = string.sub(item.abbr, 1, max_width - 1) .. "⋯"
          end

          return item
        end,
      }
    end,
  }
```

**Comment by jyuan0** [comment, marked as answer]
You could try iterating over `icons` in the `format` function and stripping the space, or redefine/override them without the space.

https://github.com/LazyVim/LazyVim/blob/main/lua/lazyvim/config/init.lua#L40

**Reply by emnnipal** [reply]
Even after removing the space on the `icons`, there is still some space. For testing purposes, I replaced the constant icon with `test` value to clearly show the issue.

<img width="253" alt="image" src="https://github.com/LazyVim/LazyVim/assets/22177776/60f56c87-28b5-42e3-8839-473fce4cd4fc">

**Reply by emnnipal** [reply]
Actually, you're right. I ended up changing the default kind icons instead. Thank you for your help!

<img width="256" alt="image" src="https://github.com/LazyVim/LazyVim/assets/22177776/a3bbe58a-98de-42bb-ae82-b870c5b39dc1">

**Reply by jyuan0** [reply]
~I think the remaining space is coming from nvim-cmp and the `fields` setting. You could try removing one of them and manually formatting the entry, but then I think there will be issues with highlighting. Kind of out of ideas at this point.~

**Reply by jyuan0** [reply]
lol, didn't see the last reply before I posted. Glad it worked for you. Good to know it works overriding the defaults.

**Reply by dpetka2001** [reply]
If you changed the defaults icons in the file mentioned above, that might lead to breakage if Lazyvim updates the icons in a future update. The easiest way I can think of would be to copy/paste all the icons in your personal spec and then in the `formatting.format` function you should `require` the icons from your personal spec.

**Reply by emnnipal** [reply]
Taking your advice into consideration, I've chosen a different approach. Rather than copying all the icons, I decided to explicitly remove the space within the formatting itself as also suggested by **jyuan0**.

```
if icons[item.kind] then
   item.kind = string.gsub(icons[item.kind], "%s+", "")
end
```

**Reply by dpetka2001** [reply]
Cool, thank you for sharing this. I must study Lua in more depth to be able to think of things like that :smile:
